kernel/src/hil/kv_system.rs

```diff
@@ -181,87 +150,88 @@ pub trait KVSystem<'a> {
             Result<(), ErrorCode>,
         ),
     >;
 
     /// Retrieves the value from a specified key.
     ///
-    /// `key`: A hashed key. This key will be used to retrieve the `value`.
-    /// `ret_buf`: A buffer to store the value to.
+    /// - `key`: A hashed key. This key will be used to retrieve the `value`.
+    /// - `ret_buf`: A buffer to store the value to.
     ///
     /// On success nothing will be returned.
     /// On error the key, ret_buf and a `Result<(), ErrorCode>` will be returned.
     ///
     /// The possible `Result<(), ErrorCode>`s are:
-    ///    `BUSY`: An operation is already in progress
-    ///    `INVAL`: An invalid parameter was passed
-    ///    `NODEVICE`: No KV store was setup
-    ///    `ENOSUPPORT`: The key could not be found.
-    ///    `SIZE`: The value is longer than the provided buffer.
+    /// - `BUSY`: An operation is already in progress
+    /// - `INVAL`: An invalid parameter was passed
+    /// - `NODEVICE`: No KV store was setup
+    /// - `ENOSUPPORT`: The key could not be found.
+    /// - `SIZE`: The value is longer than the provided buffer.
     fn get_value(
         &self,
         key: &'static mut Self::K,
-        ret_buf: &'static mut [u8],
+        ret_buf: LeasableMutableBuffer<'static, u8>,
     ) -> Result<
         (),
         (
             &'static mut Self::K,
-            &'static mut [u8],
+            LeasableMutableBuffer<'static, u8>,
             Result<(), ErrorCode>,
         ),
     >;
 
-    /// Invalidates the key in flash storage
+    /// Invalidates the key in flash storage.
     ///
-    /// `key`: A hashed key. This key will be used to remove the `value`.
+    /// - `key`: A hashed key. This key will be used to remove the `value`.
     ///
     /// On success nothing will be returned.
     /// On error the key and a `Result<(), ErrorCode>` will be returned.
     ///
     /// The possible `Result<(), ErrorCode>`s are:
-    ///    `BUSY`: An operation is already in progress
-    ///    `INVAL`: An invalid parameter was passed
-    ///    `NODEVICE`: No KV store was setup
-    ///    `ENOSUPPORT`: The key could not be found.
+    /// - `BUSY`: An operation is already in progress
+    /// - `INVAL`: An invalid parameter was passed
+    /// - `NODEVICE`: No KV store was setup
+    /// - `ENOSUPPORT`: The key could not be found.
     fn invalidate_key(
         &self,
         key: &'static mut Self::K,
     ) -> Result<(), (&'static mut Self::K, Result<(), ErrorCode>)>;
 
-    /// Perform a garbage collection on the KV Store
+    /// Perform a garbage collection on the KV Store.
     ///
-    /// For implementations that don't require garbage collecting
-    /// this can just be a NOP that returns 'Ok(0)'.
+    /// For implementations that don't require garbage collecting this should
+    /// return `Err(ErrorCode::ALREADY)`.
     ///
-    /// On success the number of bytes freed will be returned.
+    /// On success nothing will be returned.
```

Comment:
Is the justification for this change that previously, Ok(0) was a confusing result because it could mean either "no need for GC" or "I ran but didnt free any bytes"?

Reply:
Yes, and importantly, it isn't clear if you will get a callback or not.